app/uk/gov/hmrc/pillar2stubs/controllers/FinancialDataController.scala

```diff
@@ -0,0 +1,434 @@
+/*
+ * Copyright 2024 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.pillar2stubs.controllers
+
+import play.api.libs.json.Json
+import play.api.mvc.{Action, AnyContent, ControllerComponents}
+import uk.gov.hmrc.pillar2stubs.controllers.FinancialDataController._
+import uk.gov.hmrc.pillar2stubs.controllers.actions.AuthActionFilter
+import uk.gov.hmrc.pillar2stubs.models.{FinancialDataResponse, FinancialItem, FinancialTransaction}
+import uk.gov.hmrc.play.bootstrap.backend.controller.BackendController
+
+import java.time.{LocalDate, LocalDateTime}
+import javax.inject.Inject
+import scala.util.Random
+import scala.util.matching.Regex
+
+class FinancialDataController @Inject() (cc: ControllerComponents, authFilter: AuthActionFilter) extends BackendController(cc) {
+
+  def retrieveFinancialData(idType: String, idNumber: String, regimeType: String): Action[AnyContent] = (Action andThen authFilter) { _ =>
+    val lastThreeDigitsPattern: Regex = "XMPLR0000000(.{3})\\s*$".r
+
+    idNumber match {
```

Comment:
stub returns for:
400 INVALID_IDTYPE
400 INVALID_REGIMETYPE
422 INVALID_DATA
422 INVALID_REGIME